e2e/tests-dfx/pull.bash

```diff
@@ -34,4 +34,71 @@ teardown() {
     assert_match "https://example.com/e2e_project.wasm"
-}
\ No newline at end of file
+}
+
+@test "dfx pull can resolve dependencies from on-chain canister metadata" {
+    # When ran with ic-ref, got following error:
+    # Certificate is not authorized to respond to queries for this canister. While developing: Did you forget to set effective_canister_id?
+    [ "$USE_IC_REF" ] && skip "skipped for ic-ref"
+    # system-wide local replica
+    dfx_start
+
+    install_asset pullable
+
+    # 1. success path
+    ## 1.1. prepare "onchain" canisters
+    cd onchain
+
+    echo -n -e \\x00asm\\x01\\x00\\x00\\x00 > src/onchain_a/main.wasm
+
+    echo -n -e \\x00asm\\x01\\x00\\x00\\x00 > src/onchain_b/empty.wasm
+    cd src/onchain_b
+    ic-wasm empty.wasm -o main.wasm metadata "dfx:deps" -d "onchain_a:rrkah-fqaaa-aaaaa-aaaaq-cai;" -v public
+    cd ../../ # go back to root of "onchain" project
+    dfx deploy
+
+    assert_command dfx canister metadata ryjl3-tyaaa-aaaaa-aaaba-cai dfx:deps
+    assert_match "onchain_a:rrkah-fqaaa-aaaaa-aaaaq-cai;"
+
+    ## 1.2. pull onchain canisters in "app" project
+    cd ../app
+    assert_command dfx pull dep
+    assert_match "Pulling canister ryjl3-tyaaa-aaaaa-aaaba-cai...
+Pulling canister rrkah-fqaaa-aaaaa-aaaaq-cai...
+WARN: \`dfx:deps\` metadata not found in canister rrkah-fqaaa-aaaaa-aaaaq-cai."
+
+    assert_command dfx pull # if not specify canister name, all pull type canisters will be pulled
+    assert_match "Pulling canister ryjl3-tyaaa-aaaaa-aaaba-cai...
+Pulling canister rrkah-fqaaa-aaaaa-aaaaq-cai...
+WARN: \`dfx:deps\` metadata not found in canister rrkah-fqaaa-aaaaa-aaaaq-cai."
+
+    # 2. sad path: if the canister is not present on-chain
+    cd ../onchain
+    dfx canister uninstall-code onchain_a
+
+    cd ../app
+    assert_command_fail dfx pull
+    assert_contains "Failed while fetch and parse \`dfx:deps\` metadata from canister rrkah-fqaaa-aaaaa-aaaaq-cai."
+    assert_contains "Canister rrkah-fqaaa-aaaaa-aaaaq-cai has no module."
+
+    cd ../onchain
+    dfx canister stop onchain_a
+    dfx canister delete onchain_a
+
+    cd ../app
+    assert_command_fail dfx pull
+    assert_contains "Failed while fetch and parse \`dfx:deps\` metadata from canister rrkah-fqaaa-aaaaa-aaaaq-cai."
+    assert_contains "Canister rrkah-fqaaa-aaaaa-aaaaq-cai not found."
+
+    # 3. sad path: if dependency metadata cannot be read (wrong format)
+    cd ../onchain
+    cd src/onchain_b
+    ic-wasm empty.wasm -o main.wasm metadata "dfx:deps" -d "rrkah-fqaaa-aaaaa-aaaaq-cai;onchain_a" -v public
+    cd ../../ # go back to root of "onchain" project
+    dfx deploy
+
+    cd ../app
+    assert_command_fail dfx pull
+    assert_contains "Failed while fetch and parse \`dfx:deps\` metadata from canister ryjl3-tyaaa-aaaaa-aaaba-cai."
+    assert_contains "Failed to parse \`dfx:deps\` entry: rrkah-fqaaa-aaaaa-aaaaq-cai. Expected \`name:Principal\`."
+}
```

Comment:
It looks like the code handles this case, but the tests don't cover it:
- dfx project: depends on onchain_a, onchain_b
- onchain_a and onchain_b both depend on onchain_c
- dfx pull fetches dependencies for onchain_c exactly once

Reply:
It's not trivial to test it in current e2e test framework. Any idea?

Reply:
What about the e2e framework makes it difficult?  It looked like you already have in this test setup that sets metadata to indicate dependencies between canisters.

It is because log output goes to stderr?  After `assert_command_fail` "$stderr" should give you that.

Reply:
How can I assert if a line only shows once in the output?

Reply:
sort | uniq -c | grep ...